urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/controller/Controller.java

```diff
@@ -0,0 +1,346 @@
+	 */
+	public void panic() {
+		for (final CtrlServerThread cst : cServer.getServerThreads())
+		{
+			cst.writeToOutputStream(ClientCommand.PANIC);
+		}
+	}
+
+	/**
+	 * Gather client data for the controller GUI's table of connected clients.
+	 *
+	 * @param rows the table row count
+	 * @param cols the table column count
+	 * @return String[][] array with client details
+	 */
+	public String[][] gatherClientData(final int rows, final int cols) {
+
+		final ArrayList<CtrlServerThread> threadList = cServer.getServerThreads();
+		final String[][] clientData = new String[rows][cols];
+
+		for (int i = 0; i < threadList.size(); i++)
+		{
+			clientData[i][0] = threadList.get(i).getClientIP().toString();
+			clientData[i][1] = threadList.get(i).getClientID().toString();
+			clientData[i][2] = threadList.get(i).getDeployedTicketID();
+			clientData[i][3] = threadList.get(i).getClientState().toString();
+		}
+		return clientData;
+	}
+
+	/**
+	 * Check whether all clients are finished with testing
+	 */
+	public synchronized void checkForTestFinish() {
+		final ArrayList<CtrlServerThread> threadList = cServer.getServerThreads();
+
+		int clientsFinished = 0;
+
+		for (final CtrlServerThread cst : threadList)
+		{
+			clientsFinished += (cst.getClientState() == ClientState.FINISHED) ? 1 : 0;
+		}
+
+		if (threadList.size() == clientsFinished)
+		{
+			final int answer = cgui.confirmDialog("Test finished!",
+					"All clients are finished with testing. Would you like to "
+							+ "gather and merge all collected logfile data now? "
+							+ "\nPressing CANCEL will delete all gathered data and "
+							+ "reset the clients to idle state!");
+
+			if (answer == JOptionPane.OK_OPTION)
+			{
+				collectClientLogFiles();
+			} else if (answer == JOptionPane.CANCEL_OPTION)
+			{
+				panic();
+			}
+		}
+	}
+
+	/**
+	 * Collect client log files, display a progress monitor meanwhile. Happens in
+	 * a sub-thread, so that the last client connection thread which triggers
+	 * this method is free again to accept the log file!
+	 *
+	 * UPDATE: does now a recheck for the client being finished. This allows
+	 * fetching the logfiles manually.
+	 */
+	public synchronized void collectClientLogFiles() {
+
+		// foreach cst/client
+		// send sendfile command
+		// catch file in csthread (fwd it?)
+		// update pm
+		// done!
+
+		final int clientsConnected = cServer.getServerThreads().size();
+
+		final Thread thread = new Thread(new Runnable() {
+			public void run() {
+
+				int i = 1;
+
+				for (final CtrlServerThread cst : cServer.getServerThreads())
+				{
+
+					cgui.logMessage("requesting log " + i + " of " + clientsConnected);
+
+					if (cst.getClientState() == ClientState.FINISHED)
+					{
+						cst.writeToOutputStream(ClientCommand.SEND_LOGFILE);
+
+						while (cst.getClientState() != ClientState.LOGFILE_SENT)
+						{
+							try
+							{
+								Thread.sleep(500);
+							} catch (final InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)